### ISSUE TITLE
NUnit 3 Console runner executes Explicit tests when filtering by category

### ISSUE THREAD

**Issue by gilescope** [nunit/nunit#1039]
NUnit 3 Console runner executes Explicit tests when filtering by category.

We're using NUnit3 against 2.5.9 compiled test dlls, and
```
--where cat != longrunning && cat !=investigate 
```
but we're finding the Explicit tests are getting run.

Is there a way to select or not select the Explicit tests using the Test Selection Language to prevent this?
(Principle of least surprise, one would expect Explicit tests not to be run, hence raising this as an issue).

**Comment by CharliePoole** [comment]
This looks like a repeat of issue #741, which we fixed earlier, but the presence of the && operator at the highest level of the filter may introduce a new complication.

**Comment by CharliePoole** [comment]
If this can be confirmed, it should be fixed in the 3.0.1 release. I'm re-categorizing it as critical since it basically means that the && operator won't work correctly in the presence of explicit tests.

**Comment by CharliePoole** [comment]
In working this issue, I didn't initially notice that the problem was arising in running tests compiled against NUnit 2.5.9. There is a problem with 3.0, which I am fixing, but this won't really fix any issues with explicit tests under NUnit 2.5.9. Those tests are run using the NUnit V2 driver and the code for executing the tests is from NUnit 2.6.4. Once the work with 3.0 is complete, we'll have to review this WRT the legacy tests.

**Comment by CharliePoole** [comment]
Turns out that 3.0 is working just fine wrt not running explicit tests unless they are actually selected. It's a little misleading that the name of the Explicit test displays when you use --labels:All but I have confirmed that lit's not actually being executed.

That leaves only the original issue related to running NUnit V2 tests. I'm reducing the priority and marking the issue as needing confirmation.

@gilescope Can you explain how you confirmed that your Explicit tests are actually being run?

**Comment by CharliePoole** [comment]
@gilescope I am unable to confirm this. I have run NUnit V2 tests for the mock-assembly.dll under NUnit 3.0. This contains some explicit tests, which are not run. Unlike NUnit 3.0, the report does not count these tests at all, since NUnit V2 does not return information about Explicit tests - they just disappear.

Unless you can provide some reproducible steps to demonstrate this problem, I plan to close the issue.

**Comment by gilescope** [comment]
It seems it is also running 'Ignore' tests as well as Explicit tests. We can tell the tests are running as they take a long time and those tests tend to be unstable. 

It's always been a bit hit and miss as to whether a specific version of NUnit 2.x correctly understands 'Explicit'. Ideally Ignore and Explicit should just be shorthand for specific Categories - maybe by making them less like special cases.

(We've been holding off compiling against nunit 3 till ReSharper support was fully there)

**Comment by CharliePoole** [comment]
There is absolutely nothing hit and miss about How Explicit is understood by NUnit. It has worked the same way since NUnit 2.4, that is through 24 releases of V2! Resharper has a spotty history of support for Explicit. In the old days, they did not use NUnit to run NUnit tests, and so had to develop their own implementations. Once they started using nunit.core to run the tests, things settled down.

NUnit 3.0 handles is expected to handle it the same way, but there are some implentation changes, so bugs can easily arise. Some other things have changed as well, which may lead to confusion:

* NUnit never reported on Explicit tests in the past. Now it does. The console lists a count of explicit tests on the line of tests not run. Explicit tests are shown in the XML as well. That does not mean they executed.

* NUnit attributes, including Explicit, were nothing more than markers in the past. In 3.0, they actually implement the code that causes the action of the attribute, like skipping the test. Any runner that fails to call the new methods in the attributes will fail to run them correctly. NUnit V2 (nunit.core) is one such runner that will fail to run 3.0 tests. Resharper, AFAIK, uses nunit.core.

* The NUnit 3.0 runner does not know how to run NUnit V2 tests unless a special driver extension is installed. The NUnit download packages install such a driver. I don't know if Resharper plans to use it or if they will deal with 2.x and 3.0 tests separately.

As you see, I closed this issue as incomplete, since there is nothing I can reproduce. If you could provide a repro case, I'd be glad to re-open it although it would probably be a low priority if the issue only arises on a version as old as 2.5.9.

**Comment by pchalamet** [comment]
I can confirm [Explicit] tests are running when exclude filter is used on command line. If exclude filter is removed, [Explicit] tests are not tested.
Using version 3.2.1.
